### PR TITLE
Json normalization performance

### PR DESCRIPTION
**Pull request by yaauie**
## Release notes

 - Restores performance of JSON-encoding operation by deferring an encoding operation and reducing unnecessary memory copying

## What does this PR do?

 - Eliminates wasteful computation when encoding JSON

## Why is it important/What is the impact to the user?

The safety added by 8.14.1's unicode normalization came at the cost of requiring all deeply-nested strings to be encoded to utf-8 prior to being passed to jrjackson, but separate upstream fixes in jrjackson eliminated this need for _most_ cases.

This PR restores some of the lost performance while keeping identical behaviour. We can chase down the remaining performance gap (which may require a behavior change) in a separate effort.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~

## How to test this PR locally

1. export an environment variable containing deeply-nested JSON structure
   ~~~
   export MESSAGE='{"one":{"two":{"three":{"four":{"five":{"six":{"seven":{"eight":{"nine":{"ten":{"eleven":{"twelve":{"thirteen":{"fourteen":{"fifteen":{}}}}}}}}}}}}}}}}'
   ~~~

2. use that environment variable as input to a simple logstash pipeline with generator and json codec that invokes `LogStash::Json.dump`:
   ~~~
   bin/logstash -e 'input { generator { threads => 4 message => "${MESSAGE}" codec => json } } filter { ruby { code => "::LogStash::Json.dump(event.to_hash)" } } output { sink { } }'
   ~~~

3. observe throughput:
   ~~~
   watch --color '(curl --silent -XGET http://localhost:9600/_node/stats | jq --color-output "pick(.process, .flow)")'
   ~~~

   On my M3 laptop, this translated to:
   | scenario | valid utf-8 | preserved utf-8 | throughput | cpu | 
   |----------|-------------|-----------------|------------|-----|
   | no normalization (8.13.x)   | ⚠️           | ✅               | 440k eps     | 48%    |
   | pre-normalization (8.14.2)   | ✅           | ✅              | 270k eps   | 62% |
   | jackson normalization (8.14.2)   | ✅           | ⚠️              | 440k eps   | 49% |
   | filtered pre-normalization ([POC](https://gist.github.com/yaauie/c71541ae5002e69f2f0aaff8d6b20677))  | ✅           | ✅              | 320k eps   | 57% |
   | reduced copy pre-normalization (this PR)  | ✅           | ✅              | **360k eps**   | **55%** |
   
   > - `valid utf-8`:
   >   - ✅ : output is _always_ valid `UTF-8`, a pre-requisite for being valid JSON
   >   - ⚠️ : when inputs are non-`UTF-8` or invalid-`UTF-8`, output can be corrupt
   > - `preserved utf-8`:
   >   - ✅ : valid `UTF-8` sequences in binary-flagged strings are preserved
   >   - ⚠️ : each byte in a valid multibyte `UTF-8` sequences in binary-flagged strings is replaced with the unicode replacement character (lossy)

### CODE REVIEW

File: logstash-core/lib/logstash/json.rb [hunk 1]
```diff
@@ -51,7 +51,12 @@ def jruby_dump(o, options = {})
     def normalize_encoding(data)
       case data
       when String
-        LogStash::UnicodeNormalizer.normalize_string_encoding(data)
```
Comment (yaauie) [inline logstash-core/lib/logstash/json.rb:54]: we _may_ be able to delete `LogStash::UnicodeNormalizer`, as I believe it is no longer referenced.

We don't need a normalized copy of the string; we simply need to emit a string that _can_ be normalized by JrJackson. In most cases this is simply the string that we received (no copy necessary), but for strings that are flagged as `BINARY`, we need to return a copy that has been flagged as utf-8 _so that_ the `RubyString#asJavaString()` that jrjackson uses will preserve any valid utf-8 sequences that it contains.

Comment (mashhurs) [reply logstash-core/lib/logstash/json.rb:54]: - we are free to remove `LogStash::UnicodeNormalizer` as it is not referenced anywhere
- we developed unicode handling when JrJackson was v0.4.18 and with your `RubyUtils#writeRubyString` changes in in 0.4.20, you are right that we no longer need `String#encode & scrub!` _expensive_ operations

LGTM if you please remove `LogStash::UnicodeNormalizer`.

**Review by mashhurs — APPROVED** [review]
LGTM!

**Comment by yaauie** [comment]
@logstashmachine backport 8.15

**Comment by yaauie** [comment]
@logstashmachine backport 8.14

**Pull request merged**